keps/sig-api-machinery/2309-idl/syntax.md

```diff
@@ -0,0 +1,635 @@
+            containers: create-only list-map(value: Container),
+            struct Container {
+                name: string,
+                // ...
+            }
+        }
+
+        status: Status,
+
+        struct Status {
+
+        }
+    }
+
+    wrapper DNSLabel: string validates(pattern: "[a-z0-9]([-a-z0-9]*[a-z0-9])?", max-length: 64);
+
+    struct Volume {
+        name: DNSLabel,
+        source: Source,
+
+        // unions contain several variations, only one of which
+        // may be set at any given time.  By default, they have
+        // a tag of `type`.
+        // If a variant has no body, simply don't specify a tag.
+        union Source {
+            hostPath: HostPath,
+            struct HostPath {
+                path: string,
+            }
+            emptyDir: EmptyDir,
+            struct EmptyDir {}
+            // ...
+        }
+    }
+
+    // enums represent one choice of a series of constant string values
+    enum DNSPolicy {
+        /// resolve cluster first, without host networking
+        ClusterFirstWithHostNet,
+        /// resolve cluster first
+        ClusterFirst,
+        /// resolve in the default order
+        Default,
+        /// don't resolve DNS
+        None,
+    }
+
+    // markers are extensions that annotate fields or types with
+    // additional data that can be picked up by tooling.
+    // several markers are built-in (e.g. `deprecated`)
+    //
+    // The examples below imagine a marker that denotes that a given
+    // field actually contains a reference to another object.
+    // Other markers might include overrides for legacy Go type names,
+    // marking fields as being behind feature gates for automated linting
+    // and better documentation, etc.
+
+    struct SecretProjection {
+        // inline fields have their fields embedded in their parent object
+        // in systems where this is supported (e.g. JSON, Go).  In systems
+        // where not supported, generally just use the type name as the field
+        // name.  Only structs and unions may be inlined.
+
+        // "raw identifiers" are surrounded with backticks, and take two purposes
+        // the first is to allow fields and keys that have the same name as keywords
+        // (like below).
+        //
+        // The second is as an emergency "break glass" to violate the "field name must
+        // have an lowercase first letter" constraint.
+        @reference(group: "v1", version: "core", `kind`: "Secret")
+        _inline: LocalObjectReference,
+    }
+
+    @reference(same-namespace: true)
+    struct LocalObjectReference {
+        // fields are referenced using `.name`
+        // (this avoids parsing ambiguities with primitive type names)
+
+        @reference-part(part: .name)
+        name: string,
+    }
+
+}
+```
+
+```
+import (
+    types (
+       // group-versions are imported explicitly from KDL or CKDL files
+       {core/v1} from "k8s.io/api/core/v1";
```

Comment:
not sure I found that anyway: how can I add validation to existing types, i.e. specialization overlaying the existing validation?

Reply:
Wrapper sounds like this on the first sight. But I mean added validation deep in a struct, e.g. to have PodSpec embedded, but only with specially named containers.

Reply:
wrapper does surface-level changes.  There's currently no provision for deep changes like you suggest.  That being said, I could imagine something like:

(given a more complex form of field paths)

```
struct MySpec {
   template: core/v1::Pod::Spec sub-validates(path: .containers.name, pattern: "foo-.+"),
}
```

Alternatively,

```
struct MySpec {
  template: SpecialPodSpec,
}
overlay(base: Pod::Spec) SpecialPodSpec {
  containers: SpecialContainer,

  overlay(base: Pod::Spec::Container) SpecialContainer {
    name: string validates(pattern: "foo-.+"),
  }
}
```

or something similar.

Reply:
either would work for me.